SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp

```diff
@@ -378,9 +379,18 @@ void USpatialWorkerConnection::SendRemoveComponent(Worker_EntityId EntityId, Wor
 	QueueOutgoingMessage<FRemoveComponent>(EntityId, ComponentId);
 }
 
-void USpatialWorkerConnection::SendComponentUpdate(Worker_EntityId EntityId, const Worker_ComponentUpdate* ComponentUpdate)
+void USpatialWorkerConnection::SendComponentUpdate(Worker_EntityId EntityId, const Worker_ComponentUpdate* ComponentUpdate, const TraceKey Key)
 {
-	QueueOutgoingMessage<FComponentUpdate>(EntityId, *ComponentUpdate);
+#if TRACE_LIB_ACTIVE
+	if (GameInstance.IsValid())
+	{
+		if (USpatialLatencyTracer* LatencyTracer = GameInstance->GetSpatialLatencyTracer())
+		{
+			LatencyTracer->WriteToLatencyTrace(Key, TEXT("Moved update to Worker queue"));
```

Comment:
Why does this need to go in the send update function instead of being in a separate function that says we've started tracing things? Or a different function that says we are sending an update with a trace?

Reply:
Because SendComponentUpdate is called from multiple locations, and we want to catch the timing that it's handed off to the worker queue.

Reply:
It looks like there is only one place we call it with a trace key though. This currently seems to effectively be 2 methods -
if there is a trace key - start trace
send update
So for RPCs we could manually call a function to start a trace as it's already the responsibility of the caller to make sure it happens.

Reply:
We need to queue the key in the update also. So internally the worker can also add trace information. We also will want to start calling this from multiple locations. We're also logging the even of the trace being handed to onto the worker, seems like the worker has the most context about when this happens.

Reply:
My concern is that this way of structuring things is tightly coupling this system to several other systems as well as adding a load of #if statements into already hard to reason about code (not the update so much as everywhere else).

Possible alternative structure for consideration -

- All messages we want to send can have attached meta data (probably an int).
- There are hooks for special bits of work to do before sending messages (I think we need this and the above to be able to write tests anyway).
- All trace stuff is in some other module.
- This module can hook into the above mentioned stuff to do the trace work.
- If the module doesn't exist then it just doesn't register the hooks.
- If you want to send an update with a trace then you use a different send message function that starts the trace and sets the meta data key on the update (this function lives in the other module).
- The blueprint wrappers are then just wrappers for calling that module that might do nothing if the module isn't loaded.

Reply:
Discussed offline - will add a delegate for queuing and dequeing onto the worker thread